### PR TITLE
dev_1

### LINKED ISSUE #190: dev_1
Description: Creating a paneled plot showing original data and deviations from the zonal mean value

NCL Script: https://www.ncl.ucar.edu/Applications/Scripts/dev_1.ncl
NCL Plot: https://www.ncl.ucar.edu/Applications/Images/dev_1_lg.png

Category: Panel

### PR DESCRIPTION
**Pull request by hCraker**
Closes #190 
image

### CODE REVIEW

**Comment by hCraker** [comment]
@clairefio @michaelavs I'm having issues calculating the deviation from the zonal mean. I have calculated the mean using xarray, but a 1-D DataArray is returned instead of a 2-D array. I'm trying to make an array that is the same shape as the original data so I can subtract the mean from the data using an element wise operation. Do either of you know how to expand the dimension of an array so that I am left with an array where the rows are filled with the zonal mean for that longitude?

**Comment by michaelavs** [comment]
> @clairefio @michaelavs I'm having issues calculating the deviation from the zonal mean. I have calculated the mean using xarray, but a 1-D DataArray is returned instead of a 2-D array. I'm trying to make an array that is the same shape as the original data so I can subtract the mean from the data using an element wise operation. Do either of you know how to expand the dimension of an array so that I am left with an array where the rows are filled with the zonal mean for that longitude?

I'm not sure if you have already found out an answer to this, but I think you can do this with numpy ([np.reshape()](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html)). If the data is correct in your array, this should work as I think it's meant to just reformat the data array. However, if the data is missing the variable of interest, I don't know if this method will work out for you.

**Comment by hCraker** [comment]
> I'm not sure if you have already found out an answer to this, but I think you can do this with numpy ([np.reshape()](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html)). If the data is correct in your array, this should work as I think it's meant to just reformat the data array. However, if the data is missing the variable of interest, I don't know if this method will work out for you.

@michaelavs I figured out how to do it using `numpy.meshgrid` but it's not very intuitive. I tried to add clarifying comments. Please let me know if you think it would be unnecessarily difficult for a novice Python user to understand.

**Comment by michaelavs** [comment]
> > I'm not sure if you have already found out an answer to this, but I think you can do this with numpy ([np.reshape()](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html)). If the data is correct in your array, this should work as I think it's meant to just reformat the data array. However, if the data is missing the variable of interest, I don't know if this method will work out for you.
> 
> @michaelavs I figured out how to do it using `numpy.meshgrid` but it's not very intuitive. I tried to add clarifying comments. Please let me know if you think it would be unnecessarily difficult for a novice Python user to understand.

I think its a good description on why meshgrid is used and what its purpose is! Past that, my only comment before approval of this PR is just adding a couple of extra empty lines between calls/function and the tick marks on the colorbar are a different interval than the original. Other than that, great job so far!

**Comment by hCraker** [comment]
@michaelavs I just made some changes. Please let me know what you think.

File: Plots/Panels/NCL_dev_1.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+# Calculate zonal mean
+mean = TS.mean(dim='lon')
+
+# Using meshgrid, a 2-D array can be created with the same shape as the
+# temperature data with the zonal mean for each latitude filling each row.
+# This way we can subtract each element of the mean 2-D array from the
+# corresponding element in the data array.
+waste, mean_grid = np.meshgrid(TS['lon'], mean)
+# Calculate deviations from zonal mean
```
Comment (michaelavs) [inline Plots/Panels/NCL_dev_1.py:50]:
```suggestion

# Calculate deviations from zonal mean
```
Add white space between these two functions for readability

File: Plots/Panels/NCL_dev_1.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import geocat.viz.util as gvutil
+
+##############################################################################
+# Read in data:
+
+# Open a netCDF data file using xarray default engine and load the data into xarrays
+ds = xr.open_dataset(gdf.get("netcdf_files/83.nc"))
+# Extract slice of data
+TS = ds.TS.isel(time=0).drop('time')
+# Fix the artifact of not-shown-data around 0 and 360-degree longitudes
```
Comment (michaelavs) [inline Plots/Panels/NCL_dev_1.py:39]:
```suggestion

# Fix the artifact of not-shown-data around 0 and 360-degree longitudes
```
Add white space for readability

File: Plots/Panels/NCL_dev_1.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import geocat.datafiles as gdf
+from geocat.viz import cmaps as gvcmaps
+import geocat.viz.util as gvutil
+
+##############################################################################
+# Read in data:
+
+# Open a netCDF data file using xarray default engine and load the data into xarrays
+ds = xr.open_dataset(gdf.get("netcdf_files/83.nc"))
+# Extract slice of data
```
Comment (michaelavs) [inline Plots/Panels/NCL_dev_1.py:37]:
```suggestion

# Extract slice of data
```
Add white space for readability

File: Plots/Panels/NCL_dev_1.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    # Use geocat.viz.util convenience function to make plots look like NCL
+    # plots by using latitude, longitude tick labels
+    gvutil.add_lat_lon_ticklabels(ax)
+    
+    # Removing degree symbol from tick labels to more closely resemble NCL example
+    ax.yaxis.set_major_formatter(LatitudeFormatter(degree_symbol=''))
+    ax.xaxis.set_major_formatter(LongitudeFormatter(degree_symbol=''))
+
+
+# Format ticks and ticklabels for zonal average plot
```
Comment (michaelavs) [inline Plots/Panels/NCL_dev_1.py:97]: Is the "Format" comment tied to the "Use" comment?

File: Plots/Panels/NCL_dev_1.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+# Use the geocat.viz function to add minor ticks
+gvutil.add_major_minor_ticks(ax2, x_minor_per_major=2)
+
+
+# Plot original data contour lines
+contour = TS.plot.contour(ax=ax1, transform=proj, vmin=235, vmax=305,
+                          levels=np.arange(235, 305, 5), colors='black',
+                          linewidths=0.5, add_labels=False)
+# Label contours and set label backgrounds white
```
Comment (michaelavs) [inline Plots/Panels/NCL_dev_1.py:110]:
```suggestion

# Label contours and set label backgrounds white
```
Add white space fore readability

File: Plots/Panels/NCL_dev_1.py (hunk 1 above)
Comment (hCraker) [reply Plots/Panels/NCL_dev_1.py:50]: I originally grouped function calls based on what part of the plot they were manipulating or if they were both steps in a calculation. Although, I think separating them is a good call. It is easier to read.

File: Plots/Panels/NCL_dev_1.py (hunk 4 above)
Comment (hCraker) [reply Plots/Panels/NCL_dev_1.py:100]: I can combine the two comments for clarity.

File: Plots/Panels/NCL_dev_1.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                          levels=np.arange(235, 305, 5), colors='black',
+                          linewidths=0.5, add_labels=False)
+
+# Label contours and set label backgrounds white
+ax1.clabel(contour, np.arange(240, 301, 10), fmt='%d', inline=True,
+           fontsize=10)
+[txt.set_bbox(dict(facecolor='white', edgecolor='none', pad=0)) for txt in contour.labelTexts]
+
+# Add lower text box
+ax1.text(0.995, 0.02, "CONTOUR FROM 235 TO 305 BY 5",
```
Comment (anissa111) [inline Plots/Panels/NCL_dev_1.py:118]: This is a small detail, but nudging up the y position parameter from 0.02 to 0.04 here recreates the small gap under the label in the original plot

Comment (hCraker) [reply Plots/Panels/NCL_dev_1.py:118]: Done! Thanks for pointing that out.

**Comment by hCraker** [comment]
image

**Comment by hCraker** [comment]
The NCL example seems to use only the paler, central values of the `BlWhRe` colormap, but I'm not sure how to get the same colors in Python. I tried using the `geocat-viz` function `truncate_colormap` but it doesn't produce the colors I'm looking for. @jukent @michaelavs @clyne have any of you used that function before, and is it the right approach to solving this problem?

**Comment by hCraker** [comment]
> The NCL example seems to use only the paler, central values of the `BlWhRe` colormap, but I'm not sure how to get the same colors in Python. I tried using the `geocat-viz` function `truncate_colormap` but it doesn't produce the colors I'm looking for. @jukent @michaelavs @clyne have any of you used that function before, and is it the right approach to solving this problem?

I figured it out shortly after posting this question. I was using the mininum and maximum values from the data as the `minval` and `maxval` arguments instead of using values for the percentage of the colormap I want the truncated version to start and stop at.

**Comment by hCraker** [comment]
image

**Pull request merged**